README.md
hub-ctrl.c
==========
[![Travis build][travis-badge]][travis]

Control USB power on a port by port basis on some USB hubs.

Originally created NIIBE Yutaka and published to Github by Joel Dare on January
31st, 2013.

This only works on USB hubs that have the hardware necessary to allow
software controlled power switching. Most hubs DO NOT include the hardware.

Prerequisites
=============

Before compiling it, you'll need to install libusb-dev (under Ubuntu). I
installed that with the following command.

    sudo apt-get install libusb-dev

You'll also need the GCC tool chain.

Compiling
=========

To compile the hub-ctrl.c program let autotools take care of it:

    ./configure && make

That results in an executable binary called hub-ctrl.

Controlling Power
=================

You can control the power on a port using the following command.

    sudo ./hub-ctrl -h 3 -P 1 -p 0

That says to control hub 3 (-h 3) port 1 (-P 1) and to turn the power
off (-p 0). You can also use ”-p 1” to turn the power back on.

You can also specify the USB device based on the BUS and DEV numbers. Use the
following command the list the currently connected devices. It's useful to run
this with the device disconnected and then again with the device connected so
that you can tell which device is the one you are trying to target (the Targus
in my case).

    lsusb

Now that we know the BUS and DEV numbers, we can control the power using those
numbers as well. Here's the command for that.

    sudo ./hub-ctrl -b 001 -d 005 -P1 -p 0

This time we are controlling the device on BUS 001 (-b 001) device 005 (-d 005)
port 1 (-P 1) and turning the power off (-p 0).

Hubs Known to Work
==================

The following is a list of Hubs that are known to have the hardware necessary
to allow power switching.

  - D-Link-DUB-H7-High-Speed-7-Port (Old version: Silver)
  - Elecom: U2H-G4S
  - Sanwa Supply: USB-HUB14GPH
  - Targus, Inc.: PAUH212
  - Hawking Technology: UH214

Original Copyright
==================

Copyright (C) 2006 Free Software Initiative of Japan

Author: NIIBE Yutaka  <gniibe at fsij.org>

This file can be distributed under the terms and conditions of the GNU General
Public License version 2 (or later).

  [travis]: https://travis-ci.org/avionic-design/usbhubctrl
  [travis-badge]: https://travis-ci.org/avionic-design/usbhubctrl.svg?branch=master


NEWS
Release 0.6.0 (2017-03-14)
==========================
2017-03-14 Bert van Hall <bert.vanhall@avionic-design.de>
  * SW-Version 0.6.0
  * hub-ctrl:
    - use hub queries only for actual hubs and avoid breaking touchy devices
    - fix parsing for integer options

  * build:
    - add Travis CI config

Release 0.4.0 (2016-11-28)
==========================
2016-11-28 Bert van Hall <bert.vanhall@avionic-design.de>
  * SW-Version 0.4.0
  * usb_eeprom:
    - fix verification failures by adding small delays
    - switch to libusb-1.0 API
    - add function to detect EEPROMs on supported hubs

  * hub-ctrl:
    - split out and rework command line options
    - add device scanning mode
    - add version printing option
    - improve error messages
    - lots of smaller code improvements and cleanups

  * build:
    - fix missing header and linker flags
    - improve default silence
    - add and distribute udev rule

  * tests:
    - use mockup to cover libusb-1.0 dependency
    - add tests for usb_eeprom_write()
    - add tests for usb_eeprom_support()
    - add auto mode for BlueTooth test script

Release 0.2.0 (2016-11-08)
==========================

2016-11-08 Meike Vocke <meike.vocke@avionic-design.de>
  * SW-Version 0.2.0
  * Initial Release of the hub-ctrl software


tests/dummy_usb.c
#include <errno.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>

#include "dummy_usb.h"

/** USB control message Requesttype for EEPROM read */
#define USB_REQ_TYPE_READ_EEPROM	0xC0
/** USB control message Requesttype for EEPROM write */
#define USB_REQ_TYPE_WRITE_EEPROM	0x40
/** USB control message Request for read */
#define USB_REQ_READ			0x02
/** USB control message Request for write */
#define USB_REQ_WRITE			0x01

libusb_device_handle *libusb_device_handle_create()
{
	libusb_device_handle *uh = NULL;

	uh = malloc(sizeof(libusb_device_handle));
	if (!uh)
		return NULL;

	uh->msg = calloc(1, sizeof(struct usb_msg));
	if (!(uh->msg)) {
		free(uh);
		return NULL;
	}

	return uh;
}

int libusb_device_handle_free(libusb_device_handle **dev)
{
	if (!(*dev) || !((*dev)->msg))
		return -EINVAL;

	if ((*dev)->msg->bytes) {
		free((*dev)->msg->bytes);
		(*dev)->msg->bytes = NULL;
	}

	free((*dev)->msg);
	(*dev)->msg = NULL;
	free(*dev);
	*dev = NULL;

	return 0;
}

struct usb_msg *get_usb_msg(libusb_device_handle *dev)
{
	if (!dev)
		return NULL;

	return dev->msg;
}

/* declared in libusb.h */
int libusb_control_transfer(libusb_device_handle *dev_handle,
	uint8_t request_type, uint8_t bRequest, uint16_t wValue, uint16_t wIndex,
	unsigned char *data, uint16_t wLength, unsigned int timeout)
{
	if (!dev_handle || !dev_handle->msg || !data || !wLength)
		return -EINVAL;

	if (request_type != USB_REQ_TYPE_READ_EEPROM &&
			request_type != USB_REQ_TYPE_WRITE_EEPROM)
		return -ERANGE;

	if (bRequest != USB_REQ_READ && bRequest != USB_REQ_WRITE)
		return -ERANGE;

	dev_handle->msg->bytes = malloc(wLength);
	if (!dev_handle->msg->bytes)
		return -ENOMEM;

	dev_handle->msg->requesttype = request_type;
	dev_handle->msg->request = bRequest;
	dev_handle->msg->value = wValue;
	dev_handle->msg->index = wIndex;
	memcpy(dev_handle->msg->bytes, data, wLength);
	dev_handle->msg->size = wLength;
	dev_handle->msg->timeout = timeout;

	return wLength;
}

/**
 * @brief Mock for reading the device descriptor
 *
 * Mock function to imitate fetching a descriptor for a USB device. The results
 * shall be supplied via the first argument.
 *
 * @param dev Casted pointer to the desired actual descriptor
 * @param desc Pointer to the resulting descriptor
 * @return 0 on success
 * @return -EINVAL if any arguments are NULL
 * @pre dev @b must be allocated to memory at least the size of the descriptor!
 * An easy and safe way to achieve this is casting pointers around.
 */
int libusb_get_device_descriptor(libusb_device *dev,
	struct libusb_device_descriptor *desc)
{
	struct libusb_device_descriptor ld;

	if (!dev || !desc)
		return -EINVAL;

	ld = *((struct libusb_device_descriptor *)dev);
	memcpy(desc, &ld, sizeof(ld));

	return 0;
}


tests/dummy_usb.h
/**
 * @file
 * @author Meike Vocke <meike.vocke\@avionic-design.de>
 * @date 2016
 *
 * @brief libusb.h mocks for usb_eeprom functions testing
 *
 * @copyright GPLv3
 */

#ifndef DUMMY_USB_H
#define DUMMY_USB_H

#include <stdint.h>
#include <libusb.h>

/** struct for passed parameters of usb_control_msg */
struct usb_msg {
	/** usb requesttype */
	int requesttype;
	/** usb request */
	int request;
	/** pointer to a buffer */
	uint8_t *bytes;
	/** timeout in ms */
	int timeout;
	/** value for usb control message */
	int value;
	/** index für usb control message */
	int index;
	/** size of bytes in buffer */
	int size;
};

/** struct as for usb device */
struct libusb_device_handle {
	/** pointer to usb control message */
	struct usb_msg *msg;
};

/** typedef for use of libusb_device_handle as in libusb.h */
typedef struct libusb_device_handle libusb_device_handle;

/**
 * @brief Create a libusb_device_handle struct
 *
 * @return pointer to created libusb_device_handle on success
 * @return null pointer on failure
 */
libusb_device_handle *libusb_device_handle_create();

/**
 * @brief Free a libusb_device_handle struct
 *
 * @param dev pointer to libusb_device_handle
 * @return 0 on success
 * @return -errno on failure
 */
int libusb_device_handle_free(libusb_device_handle **dev);

/**
 * @brief Get usb control message from libusb_device_handle
 *
 * @param dev pointer to the usb device
 * @return usb_msg on success
 * @return null pointer when dev is not defined
 */
struct usb_msg *get_usb_msg(libusb_device_handle *dev);

#endif /* DUMMY_USB_H */


tests/check_file_io.c
#include <check.h>
#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

#include "file_io.h"

char file_name[] = "/tmp/fileXXXXXX";
uint8_t *file_buffer = NULL;
const uint8_t cmp_buffer[] = {
	0xD4, 0xB4, 0x04, 0x60, 0x65, 0x00, 0x92, 0x88,
	0x28, 0x5F, 0x00, 0x00, 0x50, 0xBE, 0x50, 0x64,
	0x32, 0x80, 0x61, 0x00, 0x02, 0x07, 0x0F, 0x0F,
	0x09, 0x04, 0x04, 0x04, 0x28, 0x00, 0x34, 0x00,
	0x4E, 0x00, 0x6E, 0x00, 0x82, 0x00, 0x82, 0x00,
	0x0C, 0x03, 0x43, 0x00, 0x31, 0x00, 0x30, 0x00,
	0x30, 0x00, 0x38, 0x00, 0x1A, 0x03, 0xFF, 0xFE,
	0xCF, 0x67, 0xEB, 0x58, 0x4A, 0x53, 0x0E, 0x5C,
	0xD4, 0x9A, 0xA1, 0x80, 0xFD, 0x4E, 0x09, 0x67,
	0x50, 0x96, 0x6C, 0x51, 0xF8, 0x53, 0x20, 0x03,
	0x41, 0x00, 0x44, 0x00, 0x54, 0x00, 0x31, 0x00,
	0x38, 0x00, 0x31, 0x00, 0x30, 0x00, 0x2D, 0x00,
	0x30, 0x00, 0x31, 0x00, 0x33, 0x00, 0x2D, 0x00,
	0x30, 0x00, 0x30, 0x00, 0x31, 0x00, 0x14, 0x03,
	0x55, 0x53, 0x00, 0x4E, 0xB3, 0x50, 0x38, 0x8F,
	0x49, 0x8F, 0x6F, 0x8B, 0xC6, 0x96, 0xBF, 0x7E,
	0x68, 0x56, 0x0E, 0x03, 0x30, 0x00, 0x30, 0x00,
	0x30, 0x00, 0x30, 0x00, 0x30, 0x00, 0x30, 0x00
};

void setup_create_file_with_data()
{
	int ret_val = 0;
	int fd;

	fd = mkstemp(file_name);
	ck_assert_int_ge(fd, 0);

	ret_val = write(fd, cmp_buffer, sizeof(cmp_buffer));
	ck_assert_int_eq(ret_val, sizeof(cmp_buffer));

	close(fd);
}

void setup_tmpfile_malloc()
{
	int fd = 0;

	fd = mkstemp(file_name);
	ck_assert_int_ge(fd, 0);

	file_buffer = malloc(sizeof(cmp_buffer));
	ck_assert_ptr_ne(file_buffer, NULL);

	close(fd);
}

void teardown()
{
	int ret_val = 0;

	ret_val = remove(file_name);
	ck_assert_int_eq(ret_val, 0);

	free(file_buffer);
	strcpy(file_name, "/tmp/fileXXXXXX");
}

START_TEST(test_file_read_boundaries)
{
	ssize_t ret_val = 0;

	ret_val = file_read(NULL, &file_buffer, sizeof(cmp_buffer));
	ck_assert_int_eq(ret_val, -EINVAL);
}
END_TEST

START_TEST(test_file_read)
{
	ssize_t ret_val = 0;

	/* read cmp_size bytes from file */
	ret_val = file_read(file_name, &file_buffer, sizeof(cmp_buffer));
	ck_assert_ptr_ne(file_buffer, NULL);
	ck_assert_int_eq(ret_val, sizeof(cmp_buffer));

	ck_assert_int_eq(memcmp(file_buffer, cmp_buffer, sizeof(cmp_buffer)), 0);

	/* read all bytes from file */
	ret_val = file_read(file_name, &file_buffer, 0);
	ck_assert_ptr_ne(file_buffer, NULL);
	ck_assert_int_eq(ret_val, sizeof(cmp_buffer));

	ck_assert_int_eq(memcmp(file_buffer, cmp_buffer, sizeof(cmp_buffer)), 0);
}
END_TEST

START_TEST(test_file_write_boundaries)
{
	int ret_val = 0;

	/* missing filename */
	ret_val = file_write(NULL, cmp_buffer, sizeof(cmp_buffer));
	ck_assert_int_eq(ret_val, -EINVAL);

	/* missing size */
	ret_val = file_write(file_name, cmp_buffer, 0);
	ck_assert_int_eq(ret_val, -EINVAL);

	/* missing buffer */
	ret_val = file_write(file_name, NULL, sizeof(cmp_buffer));
	ck_assert_int_eq(ret_val, -EINVAL);
}
END_TEST

START_TEST(test_file_write)
{
	int ret_val = 0;
	int size = 0;
	int fd;

	size = file_write(file_name, cmp_buffer, sizeof(cmp_buffer));
	ck_assert_int_eq(size, sizeof(cmp_buffer));

	fd = open(file_name, O_RDONLY);
	ck_assert_int_ge(fd, 0);

	ret_val = read(fd, file_buffer, sizeof(cmp_buffer));
	ck_assert_int_eq(ret_val, sizeof(cmp_buffer));

	ck_assert_int_eq(memcmp(file_buffer, cmp_buffer, sizeof(cmp_buffer)), 0);

	close(fd);
}
END_TEST

Suite * file_io_suite(Suite *s_file)
{
	TCase *tc_file_write;
	TCase *tc_file_read;

	tc_file_read = tcase_create("file read");
	tc_file_write = tcase_create("file write");

	tcase_add_unchecked_fixture(tc_file_read, setup_create_file_with_data, teardown);
	tcase_add_test(tc_file_read, test_file_read);
	tcase_add_test(tc_file_read, test_file_read_boundaries);

	tcase_add_unchecked_fixture(tc_file_write, setup_tmpfile_malloc,
			teardown);
	tcase_add_test(tc_file_write, test_file_write);
	tcase_add_test(tc_file_write, test_file_write_boundaries);

	suite_add_tcase(s_file, tc_file_read);
	suite_add_tcase(s_file, tc_file_write);

	return EXIT_SUCCESS;
}


hub-ctrl.c
/**
 * @mainpage Control USB power on a port by port basis on some USB hubs.
 *
 * This only works on USB hubs that have the hardware necessary to allow
 * software controlled power switching. Most hubs DO NOT include the
 * hardware.
 *
 * @author NIIBE Yutaka <gniibe at fsij.org>
 * @author Bert van Hall <bert.vanhall\@avionic-design.de>
 * @date 2006-2016
 * @copyright GPLv3
 */

#include <errno.h>
#include <usb.h>
#include <stdio.h>
#include <string.h>

#define USB_RT_HUB			(USB_TYPE_CLASS | USB_RECIP_DEVICE)
#define USB_RT_PORT			(USB_TYPE_CLASS | USB_RECIP_OTHER)
#define USB_PORT_FEAT_POWER		8
#define USB_PORT_FEAT_INDICATOR		22
#define USB_DIR_IN			0x80		/* to host */

#define COMMAND_SET_NONE		0
#define COMMAND_SET_LED			1
#define COMMAND_SET_POWER		2
#define HUB_LED_GREEN			2

#define HUB_CHAR_LPSM			0x0003
#define HUB_CHAR_PORTIND		0x0080

#define CTRL_TIMEOUT 			1000
#define USB_STATUS_SIZE 		4

#define MAX_HUBS 128

struct usb_hub_descriptor {
	unsigned char bDescLength;
	unsigned char bDescriptorType;
	unsigned char bNbrPorts;
	unsigned char wHubCharacteristics[2];
	unsigned char bPwrOn2PwrGood;
	unsigned char bHubContrCurrent;
	unsigned char data[0];
};

struct hub_info {
	int busnum;
	int devnum;
	struct usb_device *dev;
	int nport;
	int indicator_support;
};

static struct hub_info hubs[MAX_HUBS];
/** Number of hubs supporting power switching */
static int num_hubs;

static void usage(const char *progname)
{
	fprintf(stderr,
		"Usage: %s [{-h HUBNUM | -b BUSNUM -d DEVNUM}] \\\n"
		"          [-P PORT] [{-p [VALUE]|-l [VALUE]}]\n", progname);
}

static void exit_with_usage(const char *progname)
{
	usage(progname);
	exit(1);
}

static void hub_port_status(usb_dev_handle *uh, int nport)
{
	int i;

	printf(" Hub Port Status:\n");
	for (i = 0; i < nport; i++) {
		char buf[USB_STATUS_SIZE];
		int ret;

		ret = usb_control_msg (uh,
			USB_ENDPOINT_IN | USB_TYPE_CLASS | USB_RECIP_OTHER,
			USB_REQ_GET_STATUS, 0, i + 1, buf, USB_STATUS_SIZE,
			CTRL_TIMEOUT);
		if (ret < 0) {
			fprintf (stderr,
				"cannot read port %d status, %s (%d)\n",
				i + 1, strerror(errno), errno);
			break;
		}

		printf("   Port %d: %02x%02x.%02x%02x", i + 1, buf[3], buf[2],
			buf[1], buf[0]);

		printf("%s%s%s%s%s",
			(buf[2] & 0x10) ? " C_RESET" : "",
			(buf[2] & 0x08) ? " C_OC" : "",
			(buf[2] & 0x04) ? " C_SUSPEND" : "",
			(buf[2] & 0x02) ? " C_ENABLE" : "",
			(buf[2] & 0x01) ? " C_CONNECT" : "");

		printf("%s%s%s%s%s%s%s%s%s%s\n",
			(buf[1] & 0x10) ? " indicator" : "",
			(buf[1] & 0x08) ? " test" : "",
			(buf[1] & 0x04) ? " highspeed" : "",
			(buf[1] & 0x02) ? " lowspeed" : "",
			(buf[1] & 0x01) ? " power" : "",
			(buf[0] & 0x10) ? " RESET" : "",
			(buf[0] & 0x08) ? " oc" : "",
			(buf[0] & 0x04) ? " suspend" : "",
			(buf[0] & 0x02) ? " enable" : "",
			(buf[0] & 0x01) ? " connect" : "");
	}
}

static int usb_find_hubs(int listing, int verbose, int busnum, int devnum, int hub)
{
	struct usb_bus *busses;
	struct usb_bus *bus;

	num_hubs = 0;
	busses = usb_get_busses();
	if (busses == NULL) {
		perror("failed to access USB");
		return -1;
	}

	for (bus = busses; bus; bus = bus->next) {
		struct usb_device *dev;

		for (dev = bus->devices; dev; dev = dev->next) {
			usb_dev_handle *uh;
			int print = 0;

			if (dev->descriptor.bDeviceClass != USB_CLASS_HUB)
				continue;

			if (listing || (verbose && ((atoi(bus->dirname) == busnum &&
					dev->devnum == devnum) ||
					hub == num_hubs))) {
				print = 1;
			}

			uh = usb_open(dev);

			if (uh == NULL)
				continue;

			char buf[1024];
			int len;
			int nport;
			struct usb_hub_descriptor *uhd = (struct usb_hub_descriptor *)buf;
			len = usb_control_msg(uh,
				USB_DIR_IN | USB_RT_HUB, USB_REQ_GET_DESCRIPTOR,
				USB_DT_HUB << 8, 0, buf, sizeof(buf), CTRL_TIMEOUT);
			if (len <= sizeof(struct usb_hub_descriptor)) {
				perror("Can't get hub descriptor");
				usb_close(uh);
				continue;
			}

			if (!(uhd->wHubCharacteristics[0] & HUB_CHAR_PORTIND) &&
					(uhd->wHubCharacteristics[0] & HUB_CHAR_LPSM) >= 2)
				continue;

			if (print) {
				printf("Hub #%d at %s:%03d\n", num_hubs,
					bus->dirname, dev->devnum);

				switch ((uhd->wHubCharacteristics[0] & HUB_CHAR_LPSM)) {
				case 0:
					fprintf(stderr, " INFO: ganged switching.\n");
					break;
				case 1:
					fprintf(stderr, " INFO: individual power switching.\n");
					break;
				case 2:
				case 3:
					fprintf(stderr, " WARN: No power switching.\n");
					break;
				}

				if (!(uhd->wHubCharacteristics[0] & HUB_CHAR_PORTIND))
					fprintf(stderr, " WARN: Port indicators are NOT supported.\n");
			}

			nport = buf[2];
			hubs[num_hubs].busnum = atoi(bus->dirname);
			hubs[num_hubs].devnum = dev->devnum;
			hubs[num_hubs].dev = dev;
			hubs[num_hubs].indicator_support =
				(uhd->wHubCharacteristics[0] & HUB_CHAR_PORTIND) ? 1 : 0;
			hubs[num_hubs].nport = nport;

			num_hubs++;

			if (verbose)
				hub_port_status(uh, nport);

			usb_close(uh);
		}
	}

	return num_hubs;
}

int get_hub(int busnum, int devnum)
{
	int i;

	for (i = 0; i < num_hubs; i++)
		if (hubs[i].busnum == busnum && hubs[i].devnum == devnum)
			return i;

	return -1;
}

int main(int argc, const char *argv[])
{
	int busnum = 0;
	int devnum = 0;
	int cmd = COMMAND_SET_NONE;
	int port = 1;
	int value = 0;
	int request, feature, index;
	int result = 0;
	int listing = 0;
	int verbose = 0;
	int hub = -1;
	usb_dev_handle *uh = NULL;
	int i;

	if (argc == 1)
		listing = 1;

	for (i = 1; i < argc; i++) {
		if (argv[i][0] != '-')
			exit_with_usage(argv[0]);

		switch (argv[i][1]) {
		case 'h':
			if (++i >= argc || busnum > 0 || devnum > 0)
				exit_with_usage(argv[0]);
			hub = atoi(argv[i]);
			break;

		case 'b':
			if (++i >= argc || hub >= 0)
				exit_with_usage (argv[0]);
			busnum = atoi(argv[i]);
			break;

		case 'd':
			if (++i >= argc || hub >= 0)
				exit_with_usage (argv[0]);
			devnum = atoi(argv[i]);
			break;

		case 'P':
			if (++i >= argc)
				exit_with_usage (argv[0]);
			port = atoi(argv[i]);
			break;

		case 'l':
			if (cmd != COMMAND_SET_NONE)
				exit_with_usage (argv[0]);
			if (++i < argc)
				value = atoi(argv[i]);
			else
				value = HUB_LED_GREEN;
			cmd = COMMAND_SET_LED;
			break;

		case 'p':
			if (cmd != COMMAND_SET_NONE)
				exit_with_usage(argv[0]);
			if (++i < argc)
				value = atoi(argv[i]);
			else
				value= 0;
			cmd = COMMAND_SET_POWER;
			break;

		case 'v':
			verbose = 1;
			if (argc == 2)
				listing = 1;
			break;

		default:
			exit_with_usage(argv[0]);
		}
	}

	/* BUS is specified, but DEV isn't, or ... */
	if ((busnum > 0 && devnum <= 0) || (busnum <= 0 && devnum > 0))
		exit_with_usage(argv[0]);

	/* Default is the hub #0 */
	if (hub < 0 && busnum == 0)
		hub = 0;

	/* Default is POWER */
	if (cmd == COMMAND_SET_NONE)
		cmd = COMMAND_SET_POWER;

	usb_init();
	usb_find_busses();
	usb_find_devices();

	if (usb_find_hubs(listing, verbose, busnum, devnum, hub) <= 0) {
		fprintf(stderr, "No hubs found.\n");
		exit(1);
	}

	if (listing)
		exit(0);

	if (hub < 0)
		hub = get_hub(busnum, devnum);

	if (hub >= 0 && hub < num_hubs)
		uh = usb_open(hubs[hub].dev);

	if (uh == NULL) {
		fprintf(stderr, "Device not found.\n");
		exit(1);
	}

	if (cmd == COMMAND_SET_POWER) {
		if (value)
			request = USB_REQ_SET_FEATURE;
		else
			request = USB_REQ_CLEAR_FEATURE;
		feature = USB_PORT_FEAT_POWER;
		index = port;
	} else {
		request = USB_REQ_SET_FEATURE;
		feature = USB_PORT_FEAT_INDICATOR;
		index = (value << 8) | port;
		printf("port %02x value = %02x\n", port, value);
	}

	if (verbose) {
		printf("Send control message (REQUEST=%d, FEATURE=%d, INDEX=%04x)\n",
			request, feature, index);
	}

	if (usb_control_msg(uh, USB_RT_PORT, request, feature, index,
			NULL, 0, CTRL_TIMEOUT) < 0) {
		perror("failed to control.\n");
		result = 1;
	}

	if (verbose)
		hub_port_status(uh, hubs[hub].nport);

	usb_close(uh);

	exit(result);
}


src/file_io.c
#include <errno.h>
#include <fcntl.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>

#include "file_io.h"

#define CHUNK_SIZE 1024

ssize_t file_read(const char *file, uint8_t **buffer, size_t size_in)
{
	uint8_t *l_buffer = NULL;
	struct stat file_stat;
	int read_bytes = 0;
	int ret_val = 0;
	int size = 0;
	int fd;

	if (!file || !buffer)
		return -EINVAL;

	if (!strcmp(file, "-"))
		fd = STDIN_FILENO;
	else
		fd = open(file, O_RDONLY);

	if (fd < 0)
		return -errno;
	else if (fd == STDIN_FILENO && !size_in)
		return -EINVAL;

	if (!size_in) {
		ret_val = fstat(fd, &file_stat);
		if (ret_val != 0)
			return -errno;
		size = file_stat.st_size;
	} else {
		size = size_in;
	}

	if (!size) {
		ret_val = 0;
		goto cleanup;
	}

	l_buffer = malloc(size);
	if (!l_buffer) {
		ret_val = -ENOMEM;
		goto cleanup;
	}

	while (read_bytes < size) {
		ret_val = read(fd, l_buffer + read_bytes, size - read_bytes);
		if (ret_val < 0) {
			ret_val = -errno;
			free(l_buffer);
			goto cleanup;
		}
		read_bytes += ret_val;
		if (!ret_val)
			break;
	}

	ret_val = read_bytes;
	*buffer = l_buffer;

cleanup:
	if (fd > 2)
		close(fd);

	return ret_val;
}

ssize_t file_write(const char *file, const uint8_t *buffer, size_t size)
{
	int ret_val;
	int fd;

	if (!file || !buffer || !size)
		return -EINVAL;

	if (!strcmp(file, "-"))
		fd = STDOUT_FILENO;
	else
		fd = open(file, O_CREAT | O_WRONLY | O_TRUNC, 0600);
	if (fd < 0)
		return -errno;

	ret_val = write(fd, buffer, size);
	if (ret_val < 0)
		ret_val = -errno;

	if (fd > 2)
		close(fd);

	return ret_val;
}


src/usb_eeprom.c
#include <errno.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <unistd.h>
#include <libusb.h>

#include "usb_eeprom.h"

#define CYPRESS_HUB_VID		0x04b4
#define CYPRESS_HUB_PID		0x6560

int usb_eeprom_erase(libusb_device_handle *dev, size_t size)
{
	uint8_t *erase_buf;
	int ret_val = 0;

	if (!dev || !size)
		return -EINVAL;
	if (size > MAX_EEPROM_SIZE)
		return -ERANGE;

	erase_buf = malloc(size);
	if (!erase_buf)
		return -ENOMEM;
	memset(erase_buf, 0xff, size);

	ret_val = usb_eeprom_write(dev, erase_buf, size);

	free(erase_buf);

	return ret_val;
}

int usb_eeprom_read(libusb_device_handle *dev, uint8_t *buffer, size_t size)
{
	int len;

	if (!buffer || !dev || !size)
		return -EINVAL;
	if (size > MAX_EEPROM_SIZE)
		return -ERANGE;

	len = libusb_control_transfer(dev, USB_REQ_TYPE_READ_EEPROM,
		USB_REQ_READ, 0, 0, buffer, size, GET_TIMEOUT(size));

	return len;
}

int usb_eeprom_write(libusb_device_handle *dev, uint8_t *buffer, size_t size)
{
	int len;

	if (!buffer || !dev || !size)
		return -EINVAL;
	if (size > MAX_EEPROM_SIZE)
		return -ERANGE;

	len = libusb_control_transfer(dev, USB_REQ_TYPE_WRITE_EEPROM,
		USB_REQ_WRITE, 0, 0, buffer, size, GET_TIMEOUT(size));

	/*
	 * Sleep for more than 5 ms to guarantee EEPROM data is written.
	 * This is the maximal write cycle time of the EEPROM, specified
	 * in the 25AA640/25LC640 datasheet.
	 */
	usleep(5500);

	return len;
}

int usb_eeprom_support(libusb_device *dev)
{
	struct libusb_device_descriptor desc;
	int ret;

	if (!dev)
		return -EINVAL;

	ret = libusb_get_device_descriptor(dev, &desc);
	if (ret)
		return -EIO;

	if (desc.idVendor != CYPRESS_HUB_VID ||
			desc.idProduct != CYPRESS_HUB_PID)
		return ret;

	ret |= EEPROM_SUPPORT_DEVICE;

	if (desc.bDeviceClass == LIBUSB_CLASS_HUB && desc.bcdDevice != 0x9015)
		ret |= EEPROM_SUPPORT_STORAGE;
	else if (desc.bDeviceClass == LIBUSB_CLASS_VENDOR_SPEC &&
			desc.bcdDevice == 0x9015)
		ret |= EEPROM_SUPPORT_STORAGE | EEPROM_SUPPORT_BLANK;

	return ret;
}


include/file_io.h
/**
 * @file
 * @author Meike Vocke <meike.vocke\@avionic-design.de>
 * @date 2016
 *
 * @brief File reading and writing implementation
 *
 * @copyright GPLv3
 */

#ifndef FILE_IO_H
#define FILE_IO_H

#include <stdint.h>
#include <stdio.h>

/**
 * @brief Read data from file into a buffer
 *
 * Opens file and copy read data into buffer. The buffer is allocated inside
 * this function
 *
 * @param file file name, - for stdin as input
 * @param buffer pointer to the buffer pointer
 * @param size_in number of bytes to read from file/stdin, 0 reads the whole file,
 * for stdin the size_in has to be non-zero
 * @return number of actually read bytes on success
 * @return -errno on failure
 */
ssize_t file_read(const char *file, uint8_t **buffer, size_t size_in);

/**
 * @brief Write data from buffer to file
 *
 * Opens file or stdout and wirte the given buffer into the opened file
 * descriptor. If a file was opened and already exists the buffer will be
 * appended to the file. Otherwise the file was created with the given buffer
 * as data inside. In case of stdout the data was written to stdout.
 *
 * @param file file name, - for stdout as output
 * @param buffer pointer to the buffer
 * @param size size of the buffer
 * @return number of actually written bytes to file on success
 * @return -errno on failure
 */
ssize_t file_write(const char *file, const uint8_t *buffer, size_t size);

#endif


include/usb_eeprom.h
/**
 * @file
 * @author Meike Vocke <meike.vocke\@avionic-design.de>
 * @date 2016
 *
 * @brief USB hub EEPROM implementation for Cypress CY7C65620/CY7C65630
 *
 * @copyright GPLv3
 */

#ifndef USB_EEPROM_H
#define USB_EEPROM_H

#include <libusb.h>

/** USB control message Requesttype for EEPROM read */
#define USB_REQ_TYPE_READ_EEPROM	0xC0
/** USB control message Requesttype for EEPROM write */
#define USB_REQ_TYPE_WRITE_EEPROM	0x40
/** USB control message Request for read */
#define USB_REQ_READ			0x02
/** USB control message Request for write */
#define USB_REQ_WRITE			0x01
/** Time constant in ms for timeout per 256 Bytes for USB control message */
#define CTRL_TIMEOUT_PER_256_BYTES	1000
/** calculate the timeout for specific amount of bytes */
#define GET_TIMEOUT(t) (((t + 256 - 1) >> 8) * CTRL_TIMEOUT_PER_256_BYTES)
/** maximum EEPROM size */
#define MAX_EEPROM_SIZE 0x1000

/**
 * @defgroup eeprom_support_flags EEPROM support flags
 *
 * Flags indicating the level of support for EEPROM operations. Tells whether
 * an EEPROM exists, is blank or the USB device it is attached to even qualifies
 * for known EEPROM commands.
 *
 * @{
 */
#define EEPROM_SUPPORT_DEVICE		0x01	/**< USB device is qualified */
#define EEPROM_SUPPORT_STORAGE		0x02	/**< EEPROM is attached */
#define EEPROM_SUPPORT_BLANK		0x04	/**< Attached EEPROM is blank */
/** @} */

/**
 * @brief Erase EEPROM data
 *
 * Erase the data contained in the EEPROM of the USB hub.
 * This means putting 0xFF in every byte.
 *
 * @param dev pointer to the libusb_device_handle to use
 * @param size number of Bytes to erase
 * @return number of actually erased bytes on success
 * @return -errno on failure
 */
int usb_eeprom_erase(libusb_device_handle *dev, size_t size);

/**
 * @brief Read EEPROM data to buffer
 *
 * Read the data from the EEPROM of the USB hub.
 * The data will be written into the given buffer.
 *
 * @param dev pointer to the libusb_device_handle to use
 * @param buffer pointer where the data will be stored
 * @param size number of Bytes to read
 * @pre the buffer have to be allocated
 * @return number of actually read bytes on success
 * @return -errno on failure
 */
int usb_eeprom_read(libusb_device_handle *dev, uint8_t *buffer, size_t size);

/**
 * @brief Write buffer EEPROM data
 *
 * Write given data from pointer to EEPROM of the USB hub.
 *
 * @param dev pointer to the libusb_device_handle to use
 * @param buffer pointer where the data will be stored
 * @param size number of Bytes to write
 * @pre the buffer have to be allocated
 * @return number of actually written bytes on success
 * @return -errno on failure
 */
int usb_eeprom_write(libusb_device_handle *dev, uint8_t *buffer, size_t size);

/**
 * @brief Detect an attached EEPROM on a supported device
 *
 * @param dev pointer to a libusb_device
 * @return Combination of @ref eeprom_support_flags when the device is supprted
 * @return 0 for an unspported device
 * @return -EINVAL on failure if dev is @c NULL
 * @return -EIO if the descriptor cannot be read
 *
 * @note For now, only Cypress hubs (04b4:6560) are supported.
 */
int usb_eeprom_support(libusb_device *dev);

#endif /* USB_EEPROM_H */
